src/site/pages/digital/src/containers/ExprToCircuitPopup/generate.ts

```diff
@@ -1,73 +1,140 @@
+    const label = options.label ?? false;
+    const format = options.format ?? "|";
+    const ops = (format === "custom") ? (options.ops ?? Formats[0]) : Formats.find(form => form.icon === format);
 
     // Create input tokens
     const tokenList = GenerateTokens(expression, ops);
+    const action = new GroupAction([CreateDeselectAllAction(info.selections).execute()]);
     const inputMap = new Map<string, DigitalComponent>();
     for (const token of tokenList) {
         if (token.type !== "input" || inputMap.has(token.name))
             continue;
         inputMap.set(token.name, Create<DigitalComponent>(input));
-        inputMap.get(token.name).setName(token.name);
+        action.add(new SetNameAction(inputMap.get(token.name), token.name).execute());
     }
 
     // Create output LED
-    const o = new LED();
-    o.setName("Output");
+    const o = Create<DigitalComponent>(output);
+    action.add(new SetNameAction(o, "Output").execute());
 
     // Get the generated circuit
-    const circuit = ExpressionToCircuit(inputMap, expression, o, ops);
+    let circuit = new DigitalObjectSet();
+    try {
+        circuit = ExpressionToCircuit(inputMap, expression, o, ops);
+    } catch (err) {
+        action.undo(); // Undo any actions that have been done so far
+        throw err;
+    }
+
+    action.add(new AddGroupAction(info.designer, circuit).execute());
 
     // Get the location of the top left corner of the screen, the 1.5 acts as a modifier
     //  so that the components are not literally in the uppermost leftmost corner
     const startPos = info.camera.getPos().sub(info.camera.getCenter().scale(info.camera.getZoom()/1.5));
-    const action = new GroupAction([CreateDeselectAllAction(info.selections).execute(),
-                                    new AddGroupAction(info.designer, circuit).execute()]);
-    const [negateAction, negated] = CreateNegatedGatesAction(info.designer, circuit);
-    action.add(negateAction);
-    OrganizeMinDepth(negated, startPos);
+    // TODO: Replace with a better (action based) way of organizing a circuit
+    OrganizeMinDepth(circuit, startPos);
+
+    const circuitComponents = circuit.getComponents();
+
+    // Add labels next to inputs
+    // TODO: This will have to be redone when there is a better organization algorithm
+    if (label) {
+        for (const [name, component] of inputMap) {
+            const newLabel = Create<Label>("Label");
+            // 100 is an arbitrary value here
+            const pos = component.getPos().sub(100, 0);
+            action.add(new PlaceAction(info.designer, newLabel).execute());
+            action.add(new SetNameAction(newLabel, name).execute());
+            action.add(new TranslateAction([newLabel], [newLabel.getPos()], [pos]).execute());
+            circuitComponents.push(newLabel);
+        }
+    }
+
+    // Set clock frequencies, also connect to oscilloscope if that option is set
+    if (input === "Clock") {
```

Comment:
I think all these post-process if-statements should be moved into their own function